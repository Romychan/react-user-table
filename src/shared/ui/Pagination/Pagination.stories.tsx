import type { Meta, StoryObj } from '@storybook/react';
import { useArgs } from '@storybook/preview-api';
import { action } from '@storybook/addon-actions';

import { Pagination } from './Pagination';

const meta: Meta<typeof Pagination> = {
  title: 'UI/Pagination',
  component: Pagination,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: { onPageChange: { action: 'onPageChange' } },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    currentPage: 1,
    totalCount: 120,
  },
  render: function Render(args) {
    const [{ currentPage }, updateArgs] = useArgs();

    const handlePageChange = (page: number) => {
      action('onPageChange')(page);
      updateArgs({ currentPage: page });
    };

    return (
      <Pagination
        {...args}
        currentPage={currentPage}
        onPageChange={handlePageChange}
      />
    );
  },
};
