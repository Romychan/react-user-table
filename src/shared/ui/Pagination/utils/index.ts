/**
 * A function for creating an array of pages in a specified range
 *
 * @param start The start of the range
 * @param end The end of the range
 *
 * @returns List of page numbers
 */
export const generatePagesArray = (start: number, end: number) => {
  return Array.from({ length: end - start }, (_, index) => index + start);
};

/**
 * A function for calculating the range of displayed pages
 *
 * @param currentPage Current page number
 * @param totalCount Total number of items
 * @param siblingPage Number of pages displayed before and after the current page
 *
 * @returns List of page numbers
 */
export const getPagesRange = (
  currentPage: number,
  totalCount: number,
  siblingPage: number,
) => {
  const totalSiblingPage = 2 * siblingPage + 1;
  const floorSiblingPage = Math.floor(totalSiblingPage / 2);
  const ceilSiblingPage = Math.ceil(totalSiblingPage / 2);

  if (totalCount < totalSiblingPage) {
    return generatePagesArray(1, totalCount + 1);
  } else if (currentPage >= 1 && currentPage <= ceilSiblingPage) {
    return generatePagesArray(1, totalSiblingPage + 1);
  } else if (currentPage + floorSiblingPage >= totalCount) {
    return generatePagesArray(
      totalCount - totalSiblingPage + 1,
      totalCount + 1,
    );
  } else {
    return generatePagesArray(
      currentPage - ceilSiblingPage + 1,
      currentPage + floorSiblingPage + 1,
    );
  }
};
