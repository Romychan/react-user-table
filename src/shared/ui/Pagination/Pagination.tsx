import { useMemo } from 'react';

import { cl } from '../../lib/utils';
import { Button } from '../Button';
import { Icon } from '../Icon';

import { getPagesRange } from './utils';
import styles from './Pagination.module.scss';

interface IPaginationProps {
  /** Controlled current page */
  currentPage: number;
  /** Additional class for the component */
  className?: string;
  /** Total number of items
   *
   * @default 1
   */
  totalCount?: number;
  /**
   * Number of items per page
   *
   * @default 10
   */
  pageSize?: number;
  /**
   * The number of pages displayed before and after the current page
   *
   * @default 2
   */
  siblingPage?: number;
  /**
   * The callback that will be called when the page changes
   *
   * @param page The selected page
   */
  onPageChange: (page: number) => void;
}

/** This component is used to display page numbering and navigation between pages */
export const Pagination = ({
  currentPage,
  onPageChange,
  totalCount = 1,
  pageSize = 10,
  siblingPage = 2,
  className = '',
}: IPaginationProps) => {
  const totalCountPage = Math.ceil(totalCount / pageSize);
  const pagesRange = useMemo(
    () => getPagesRange(currentPage, totalCountPage, siblingPage),
    [currentPage, totalCountPage, siblingPage],
  );

  return (
    <div className={cl(styles.container, className)} data-testid="pagination">
      <Button
        variant="secondary"
        size="sm"
        data-testid="pagination-prev"
        disabled={currentPage <= 1}
        onClick={() => onPageChange(currentPage - 1)}
        className={cl(styles.page, styles.prev)}
        isIconOnly
      >
        <Icon name="chevron" />
      </Button>

      {pagesRange[0] > 1 ? (
        <Button
          variant="secondary"
          size="sm"
          data-testid="pagination-prev-dot"
          onClick={() => onPageChange(currentPage - 1)}
          className={styles.page}
          isIconOnly
        >
          ...
        </Button>
      ) : null}

      {pagesRange.map((number) => (
        <Button
          key={number}
          variant={currentPage == number ? 'primary' : 'secondary'}
          size="sm"
          onClick={() => onPageChange(number)}
          className={styles.page}
          isIconOnly
        >
          {number}
        </Button>
      ))}

      {totalCountPage !== pagesRange[pagesRange.length - 1] ? (
        <Button
          variant="secondary"
          size="sm"
          data-testid="pagination-next-dot"
          onClick={() => onPageChange(currentPage + 1)}
          className={styles.page}
          isIconOnly
        >
          ...
        </Button>
      ) : null}

      <Button
        variant="secondary"
        size="sm"
        data-testid="pagination-next"
        disabled={currentPage >= totalCountPage}
        onClick={() => onPageChange(currentPage + 1)}
        className={cl(styles.page, styles.next)}
        isIconOnly
      >
        <Icon name="chevron" />
      </Button>
    </div>
  );
};
