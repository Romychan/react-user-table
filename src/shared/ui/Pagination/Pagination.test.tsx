import { fireEvent, screen, render } from '@testing-library/react';
import { vi } from 'vitest';

import { Pagination } from './Pagination';

describe('Pagination', () => {
  it('should render correctly', () => {
    const handleClick = vi.fn();

    render(
      <Pagination
        currentPage={1}
        onPageChange={handleClick}
        totalCount={100}
      />,
    );
    const paginationElement = screen.getByTestId('pagination');
    const ellipsisElement = screen.getByTestId('pagination-next-dot');

    expect(paginationElement).toBeInTheDocument();
    expect(ellipsisElement).toBeInTheDocument();
  });

  it('should be disable prev button if the first page is selected', () => {
    const handleClick = vi.fn();

    render(
      <Pagination currentPage={1} onPageChange={handleClick} totalCount={20} />,
    );

    const prevButtonElement = screen.getByTestId('pagination-prev');

    expect(prevButtonElement).toHaveAttribute('disabled');
  });

  it('should disable next button if the last page is selected', () => {
    const handleClick = vi.fn();

    render(
      <Pagination currentPage={2} onPageChange={handleClick} totalCount={20} />,
    );

    const nextButtonElement = screen.getByTestId('pagination-next');

    expect(nextButtonElement).toHaveAttribute('disabled');
  });

  it('should called handleClick after click prev button', () => {
    const handleClick = vi.fn();

    render(
      <Pagination
        currentPage={6}
        onPageChange={handleClick}
        totalCount={100}
      />,
    );

    const prevButtonElement = screen.getByTestId('pagination-prev');

    fireEvent.click(prevButtonElement);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('should called handleClick after click next button', () => {
    const handleClick = vi.fn();

    render(
      <Pagination
        currentPage={6}
        onPageChange={handleClick}
        totalCount={100}
      />,
    );

    const nextButtonElement = screen.getByTestId('pagination-next');

    fireEvent.click(nextButtonElement);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot', () => {
    const handleClick = vi.fn();
    const container = render(
      <Pagination
        currentPage={6}
        onPageChange={handleClick}
        totalCount={100}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
