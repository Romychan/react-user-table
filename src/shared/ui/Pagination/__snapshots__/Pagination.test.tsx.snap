// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_900fc1"
        data-testid="pagination"
      >
        <button
          class="_button_0302f3 _page_900fc1 _prev_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
          data-testid="pagination-prev"
        >
          <svg
            class="icon icon-chevron"
            data-testid="icon"
            height="16"
            width="16"
          >
            <use
              xlink:href="/images/icons.svg#chevron"
            />
          </svg>
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
          data-testid="pagination-prev-dot"
        >
          ...
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        >
          4
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        >
          5
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _primary_0302f3 _sm_0302f3 _icon_0302f3"
        >
          6
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        >
          7
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        >
          8
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
          data-testid="pagination-next-dot"
        >
          ...
        </button>
        <button
          class="_button_0302f3 _page_900fc1 _next_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
          data-testid="pagination-next"
        >
          <svg
            class="icon icon-chevron"
            data-testid="icon"
            height="16"
            width="16"
          >
            <use
              xlink:href="/images/icons.svg#chevron"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_900fc1"
      data-testid="pagination"
    >
      <button
        class="_button_0302f3 _page_900fc1 _prev_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        data-testid="pagination-prev"
      >
        <svg
          class="icon icon-chevron"
          data-testid="icon"
          height="16"
          width="16"
        >
          <use
            xlink:href="/images/icons.svg#chevron"
          />
        </svg>
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        data-testid="pagination-prev-dot"
      >
        ...
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
      >
        4
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
      >
        5
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _primary_0302f3 _sm_0302f3 _icon_0302f3"
      >
        6
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
      >
        7
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
      >
        8
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        data-testid="pagination-next-dot"
      >
        ...
      </button>
      <button
        class="_button_0302f3 _page_900fc1 _next_900fc1 _secondary_0302f3 _sm_0302f3 _icon_0302f3"
        data-testid="pagination-next"
      >
        <svg
          class="icon icon-chevron"
          data-testid="icon"
          height="16"
          width="16"
        >
          <use
            xlink:href="/images/icons.svg#chevron"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
